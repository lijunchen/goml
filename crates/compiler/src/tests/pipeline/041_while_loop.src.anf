fn sum_to(limit/0: int) -> int {
  let acc/1 = ref(0) in
  let i/2 = ref(0) in
  let t8 = while let t9 = ref_get(i/2) in
  int_less(t9, limit/0) {
    let current/3 = ref_get(i/2) in
    let t11 = ref_get(acc/1) in
    let t10 = int_add(t11, current/3) in
    let mtmp0 = ref_set(acc/1, t10) in
    let t12 = int_add(current/3, 1) in
    let mtmp1 = ref_set(i/2, t12) in
    ()
  } in
  let mtmp2 = (t8) in
  ref_get(acc/1)
}

fn sum_even(limit/4: int) -> int {
  let acc/5 = ref(0) in
  let i/6 = ref(0) in
  let is_even/7 = ref(true) in
  let t13 = while let t14 = ref_get(i/6) in
  int_less(t14, limit/4) {
    let current/8 = ref_get(i/6) in
    let t15 = int_add(current/8, 1) in
    let mtmp3 = ref_set(i/6, t15) in
    let add_now/9 = ref_get(is_even/7) in
    let t16 = bool_not(add_now/9) in
    let mtmp4 = ref_set(is_even/7, t16) in
    if add_now/9 {
      let t18 = ref_get(acc/5) in
      let t17 = int_add(t18, current/8) in
      ref_set(acc/5, t17)
    } else {
      ()
    }
  } in
  let mtmp5 = (t13) in
  ref_get(acc/5)
}

fn main() -> unit {
  let first/10 = sum_to(5) in
  let evens/11 = sum_even(6) in
  let t20 = int_to_string(first/10) in
  let t19 = string_add("sum_to(5)=", t20) in
  let mtmp6 = string_println(t19) in
  let t22 = int_to_string(evens/11) in
  let t21 = string_add("sum_even(6)=", t22) in
  let mtmp7 = string_println(t21) in
  ()
}