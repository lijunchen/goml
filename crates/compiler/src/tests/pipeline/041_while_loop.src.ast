fn sum_to(limit: int) -> int {
    let acc = ref(0) in
    let i = ref(0) in
    let _ = (while int_less(ref_get(i), limit) {
        let current = ref_get(i) in
        let _ = ref_set(acc, ref_get(acc) + current) in
        let _ = ref_set(i, current + 1) in
        ()
    }) in
    ref_get(acc)
}

fn sum_even(limit: int) -> int {
    let acc = ref(0) in
    let i = ref(0) in
    let is_even = ref(true) in
    let _ = (while int_less(ref_get(i), limit) {
        let current = ref_get(i) in
        let _ = ref_set(i, current + 1) in
        let add_now = ref_get(is_even) in
        let _ = ref_set(is_even, !add_now) in
        if add_now {
            ref_set(acc, ref_get(acc) + current)
        } else {
            ()
        }
    }) in
    ref_get(acc)
}

fn main() {
    let first = sum_to(5) in
    let evens = sum_even(6) in
    let _ = string_println(string_add("sum_to(5)=", int_to_string(first))) in
    let _ = string_println(string_add("sum_even(6)=", int_to_string(evens))) in
    ()
}

