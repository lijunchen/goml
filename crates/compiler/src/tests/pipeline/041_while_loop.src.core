fn sum_to(limit/0: int) -> int {
  let acc/1 = ref(0) in
  let i/2 = ref(0) in
  let mtmp2 = (while int_less(ref_get(i/2), limit/0) {
      let current/3 = ref_get(i/2) in
      let mtmp0 = ref_set(acc/1, int_add(ref_get(acc/1), current/3)) in
      let mtmp1 = ref_set(i/2, int_add(current/3, 1)) in
      ()
  }) in
  ref_get(acc/1)
}

fn sum_even(limit/4: int) -> int {
  let acc/5 = ref(0) in
  let i/6 = ref(0) in
  let is_even/7 = ref(true) in
  let mtmp5 = (while int_less(ref_get(i/6), limit/4) {
      let current/8 = ref_get(i/6) in
      let mtmp3 = ref_set(i/6, int_add(current/8, 1)) in
      let add_now/9 = ref_get(is_even/7) in
      let mtmp4 = ref_set(is_even/7, bool_not(add_now/9)) in
      if add_now/9 {
          ref_set(acc/5, int_add(ref_get(acc/5), current/8))
      } else {
          ()
      }
  }) in
  ref_get(acc/5)
}

fn main() -> unit {
  let first/10 = sum_to(5) in
  let evens/11 = sum_even(6) in
  let mtmp6 = string_println(string_add("sum_to(5)=", int_to_string(first/10))) in
  let mtmp7 = string_println(string_add("sum_even(6)=", int_to_string(evens/11))) in
  ()
}