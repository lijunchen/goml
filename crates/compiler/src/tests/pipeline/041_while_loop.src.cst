FILE@0..921
  FN@0..276
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..9 "sum_to"
    PARAM_LIST@9..22
      LParen@9..10 "("
      PARAM@10..20
        Lident@10..15 "limit"
        Colon@15..16 ":"
        Whitespace@16..17 " "
        TYPE_INT@17..20
          IntKeyword@17..20 "int"
      RParen@20..21 ")"
      Whitespace@21..22 " "
    Arrow@22..24 "->"
    Whitespace@24..25 " "
    TYPE_INT@25..29
      IntKeyword@25..28 "int"
      Whitespace@28..29 " "
    BLOCK@29..276
      LBrace@29..30 "{"
      Whitespace@30..33 "\n  "
      EXPR_LET@33..273
        LetKeyword@33..36 "let"
        Whitespace@36..37 " "
        PATTERN_VARIABLE@37..41
          Lident@37..40 "acc"
          Whitespace@40..41 " "
        Eq@41..42 "="
        Whitespace@42..43 " "
        EXPR_LET_VALUE@43..50
          EXPR_CALL@43..50
            EXPR_LIDENT@43..46
              Lident@43..46 "ref"
            ARG_LIST@46..50
              LParen@46..47 "("
              ARG@47..48
                EXPR_INT@47..48
                  Int@47..48 "0"
              RParen@48..49 ")"
              Whitespace@49..50 " "
        InKeyword@50..52 "in"
        Whitespace@52..55 "\n  "
        EXPR_LET_BODY@55..273
          EXPR_LET@55..273
            LetKeyword@55..58 "let"
            Whitespace@58..59 " "
            PATTERN_VARIABLE@59..61
              Lident@59..60 "i"
              Whitespace@60..61 " "
            Eq@61..62 "="
            Whitespace@62..63 " "
            EXPR_LET_VALUE@63..70
              EXPR_CALL@63..70
                EXPR_LIDENT@63..66
                  Lident@63..66 "ref"
                ARG_LIST@66..70
                  LParen@66..67 "("
                  ARG@67..68
                    EXPR_INT@67..68
                      Int@67..68 "0"
                  RParen@68..69 ")"
                  Whitespace@69..70 " "
            InKeyword@70..72 "in"
            Whitespace@72..75 "\n  "
            EXPR_LET_BODY@75..273
              EXPR_LET@75..273
                LetKeyword@75..78 "let"
                Whitespace@78..79 " "
                PATTERN_WILDCARD@79..81
                  WildcardKeyword@79..80 "_"
                  Whitespace@80..81 " "
                Eq@81..82 "="
                Whitespace@82..83 " "
                EXPR_LET_VALUE@83..255
                  EXPR_TUPLE@83..255
                    LParen@83..84 "("
                    EXPR_WHILE@84..253
                      WhileKeyword@84..89 "while"
                      Whitespace@89..90 " "
                      EXPR_WHILE_COND@90..118
                        EXPR_CALL@90..118
                          EXPR_LIDENT@90..98
                            Lident@90..98 "int_less"
                          ARG_LIST@98..118
                            LParen@98..99 "("
                            ARG@99..111
                              EXPR_CALL@99..109
                                EXPR_LIDENT@99..106
                                  Lident@99..106 "ref_get"
                                ARG_LIST@106..109
                                  LParen@106..107 "("
                                  ARG@107..108
                                    EXPR_LIDENT@107..108
                                      Lident@107..108 "i"
                                  RParen@108..109 ")"
                              Comma@109..110 ","
                              Whitespace@110..111 " "
                            ARG@111..116
                              EXPR_LIDENT@111..116
                                Lident@111..116 "limit"
                            RParen@116..117 ")"
                            Whitespace@117..118 " "
                      EXPR_WHILE_BODY@118..253
                        BLOCK@118..253
                          LBrace@118..119 "{"
                          Whitespace@119..124 "\n    "
                          EXPR_LET@124..252
                            LetKeyword@124..127 "let"
                            Whitespace@127..128 " "
                            PATTERN_VARIABLE@128..136
                              Lident@128..135 "current"
                              Whitespace@135..136 " "
                            Eq@136..137 "="
                            Whitespace@137..138 " "
                            EXPR_LET_VALUE@138..149
                              EXPR_CALL@138..149
                                EXPR_LIDENT@138..145
                                  Lident@138..145 "ref_get"
                                ARG_LIST@145..149
                                  LParen@145..146 "("
                                  ARG@146..147
                                    EXPR_LIDENT@146..147
                                      Lident@146..147 "i"
                                  RParen@147..148 ")"
                                  Whitespace@148..149 " "
                            InKeyword@149..151 "in"
                            Whitespace@151..156 "\n    "
                            EXPR_LET_BODY@156..252
                              EXPR_LET@156..252
                                LetKeyword@156..159 "let"
                                Whitespace@159..160 " "
                                PATTERN_WILDCARD@160..162
                                  WildcardKeyword@160..161 "_"
                                  Whitespace@161..162 " "
                                Eq@162..163 "="
                                Whitespace@163..164 " "
                                EXPR_LET_VALUE@164..201
                                  EXPR_CALL@164..201
                                    EXPR_LIDENT@164..171
                                      Lident@164..171 "ref_set"
                                    ARG_LIST@171..201
                                      LParen@171..172 "("
                                      ARG@172..177
                                        EXPR_LIDENT@172..175
                                          Lident@172..175 "acc"
                                        Comma@175..176 ","
                                        Whitespace@176..177 " "
                                      ARG@177..199
                                        EXPR_BINARY@177..199
                                          EXPR_CALL@177..190
                                            EXPR_LIDENT@177..184
                                              Lident@177..184 "ref_get"
                                            ARG_LIST@184..190
                                              LParen@184..185 "("
                                              ARG@185..188
                                                EXPR_LIDENT@185..188
                                                  Lident@185..188 "acc"
                                              RParen@188..189 ")"
                                              Whitespace@189..190 " "
                                          Plus@190..191 "+"
                                          Whitespace@191..192 " "
                                          EXPR_LIDENT@192..199
                                            Lident@192..199 "current"
                                      RParen@199..200 ")"
                                      Whitespace@200..201 " "
                                InKeyword@201..203 "in"
                                Whitespace@203..208 "\n    "
                                EXPR_LET_BODY@208..252
                                  EXPR_LET@208..252
                                    LetKeyword@208..211 "let"
                                    Whitespace@211..212 " "
                                    PATTERN_WILDCARD@212..214
                                      WildcardKeyword@212..213 "_"
                                      Whitespace@213..214 " "
                                    Eq@214..215 "="
                                    Whitespace@215..216 " "
                                    EXPR_LET_VALUE@216..240
                                      EXPR_CALL@216..240
                                        EXPR_LIDENT@216..223
                                          Lident@216..223 "ref_set"
                                        ARG_LIST@223..240
                                          LParen@223..224 "("
                                          ARG@224..227
                                            EXPR_LIDENT@224..225
                                              Lident@224..225 "i"
                                            Comma@225..226 ","
                                            Whitespace@226..227 " "
                                          ARG@227..238
                                            EXPR_BINARY@227..238
                                              EXPR_LIDENT@227..235
                                                Lident@227..234 "current"
                                                Whitespace@234..235 " "
                                              Plus@235..236 "+"
                                              Whitespace@236..237 " "
                                              EXPR_INT@237..238
                                                Int@237..238 "1"
                                          RParen@238..239 ")"
                                          Whitespace@239..240 " "
                                    InKeyword@240..242 "in"
                                    Whitespace@242..247 "\n    "
                                    EXPR_LET_BODY@247..252
                                      EXPR_UNIT@247..252
                                        LParen@247..248 "("
                                        RParen@248..249 ")"
                                        Whitespace@249..252 "\n  "
                          RBrace@252..253 "}"
                    RParen@253..254 ")"
                    Whitespace@254..255 " "
                InKeyword@255..257 "in"
                Whitespace@257..260 "\n  "
                EXPR_LET_BODY@260..273
                  EXPR_CALL@260..273
                    EXPR_LIDENT@260..267
                      Lident@260..267 "ref_get"
                    ARG_LIST@267..273
                      LParen@267..268 "("
                      ARG@268..271
                        EXPR_LIDENT@268..271
                          Lident@268..271 "acc"
                      RParen@271..272 ")"
                      Whitespace@272..273 "\n"
      RBrace@273..274 "}"
      Whitespace@274..276 "\n\n"
  FN@276..692
    FnKeyword@276..278 "fn"
    Whitespace@278..279 " "
    Lident@279..287 "sum_even"
    PARAM_LIST@287..300
      LParen@287..288 "("
      PARAM@288..298
        Lident@288..293 "limit"
        Colon@293..294 ":"
        Whitespace@294..295 " "
        TYPE_INT@295..298
          IntKeyword@295..298 "int"
      RParen@298..299 ")"
      Whitespace@299..300 " "
    Arrow@300..302 "->"
    Whitespace@302..303 " "
    TYPE_INT@303..307
      IntKeyword@303..306 "int"
      Whitespace@306..307 " "
    BLOCK@307..692
      LBrace@307..308 "{"
      Whitespace@308..311 "\n  "
      EXPR_LET@311..689
        LetKeyword@311..314 "let"
        Whitespace@314..315 " "
        PATTERN_VARIABLE@315..319
          Lident@315..318 "acc"
          Whitespace@318..319 " "
        Eq@319..320 "="
        Whitespace@320..321 " "
        EXPR_LET_VALUE@321..328
          EXPR_CALL@321..328
            EXPR_LIDENT@321..324
              Lident@321..324 "ref"
            ARG_LIST@324..328
              LParen@324..325 "("
              ARG@325..326
                EXPR_INT@325..326
                  Int@325..326 "0"
              RParen@326..327 ")"
              Whitespace@327..328 " "
        InKeyword@328..330 "in"
        Whitespace@330..333 "\n  "
        EXPR_LET_BODY@333..689
          EXPR_LET@333..689
            LetKeyword@333..336 "let"
            Whitespace@336..337 " "
            PATTERN_VARIABLE@337..339
              Lident@337..338 "i"
              Whitespace@338..339 " "
            Eq@339..340 "="
            Whitespace@340..341 " "
            EXPR_LET_VALUE@341..348
              EXPR_CALL@341..348
                EXPR_LIDENT@341..344
                  Lident@341..344 "ref"
                ARG_LIST@344..348
                  LParen@344..345 "("
                  ARG@345..346
                    EXPR_INT@345..346
                      Int@345..346 "0"
                  RParen@346..347 ")"
                  Whitespace@347..348 " "
            InKeyword@348..350 "in"
            Whitespace@350..353 "\n  "
            EXPR_LET_BODY@353..689
              EXPR_LET@353..689
                LetKeyword@353..356 "let"
                Whitespace@356..357 " "
                PATTERN_VARIABLE@357..365
                  Lident@357..364 "is_even"
                  Whitespace@364..365 " "
                Eq@365..366 "="
                Whitespace@366..367 " "
                EXPR_LET_VALUE@367..377
                  EXPR_CALL@367..377
                    EXPR_LIDENT@367..370
                      Lident@367..370 "ref"
                    ARG_LIST@370..377
                      LParen@370..371 "("
                      ARG@371..375
                        EXPR_BOOL@371..375
                          TrueKeyword@371..375 "true"
                      RParen@375..376 ")"
                      Whitespace@376..377 " "
                InKeyword@377..379 "in"
                Whitespace@379..382 "\n  "
                EXPR_LET_BODY@382..689
                  EXPR_LET@382..689
                    LetKeyword@382..385 "let"
                    Whitespace@385..386 " "
                    PATTERN_WILDCARD@386..388
                      WildcardKeyword@386..387 "_"
                      Whitespace@387..388 " "
                    Eq@388..389 "="
                    Whitespace@389..390 " "
                    EXPR_LET_VALUE@390..671
                      EXPR_TUPLE@390..671
                        LParen@390..391 "("
                        EXPR_WHILE@391..669
                          WhileKeyword@391..396 "while"
                          Whitespace@396..397 " "
                          EXPR_WHILE_COND@397..425
                            EXPR_CALL@397..425
                              EXPR_LIDENT@397..405
                                Lident@397..405 "int_less"
                              ARG_LIST@405..425
                                LParen@405..406 "("
                                ARG@406..418
                                  EXPR_CALL@406..416
                                    EXPR_LIDENT@406..413
                                      Lident@406..413 "ref_get"
                                    ARG_LIST@413..416
                                      LParen@413..414 "("
                                      ARG@414..415
                                        EXPR_LIDENT@414..415
                                          Lident@414..415 "i"
                                      RParen@415..416 ")"
                                  Comma@416..417 ","
                                  Whitespace@417..418 " "
                                ARG@418..423
                                  EXPR_LIDENT@418..423
                                    Lident@418..423 "limit"
                                RParen@423..424 ")"
                                Whitespace@424..425 " "
                          EXPR_WHILE_BODY@425..669
                            BLOCK@425..669
                              LBrace@425..426 "{"
                              Whitespace@426..431 "\n    "
                              EXPR_LET@431..668
                                LetKeyword@431..434 "let"
                                Whitespace@434..435 " "
                                PATTERN_VARIABLE@435..443
                                  Lident@435..442 "current"
                                  Whitespace@442..443 " "
                                Eq@443..444 "="
                                Whitespace@444..445 " "
                                EXPR_LET_VALUE@445..456
                                  EXPR_CALL@445..456
                                    EXPR_LIDENT@445..452
                                      Lident@445..452 "ref_get"
                                    ARG_LIST@452..456
                                      LParen@452..453 "("
                                      ARG@453..454
                                        EXPR_LIDENT@453..454
                                          Lident@453..454 "i"
                                      RParen@454..455 ")"
                                      Whitespace@455..456 " "
                                InKeyword@456..458 "in"
                                Whitespace@458..463 "\n    "
                                EXPR_LET_BODY@463..668
                                  EXPR_LET@463..668
                                    LetKeyword@463..466 "let"
                                    Whitespace@466..467 " "
                                    PATTERN_WILDCARD@467..469
                                      WildcardKeyword@467..468 "_"
                                      Whitespace@468..469 " "
                                    Eq@469..470 "="
                                    Whitespace@470..471 " "
                                    EXPR_LET_VALUE@471..495
                                      EXPR_CALL@471..495
                                        EXPR_LIDENT@471..478
                                          Lident@471..478 "ref_set"
                                        ARG_LIST@478..495
                                          LParen@478..479 "("
                                          ARG@479..482
                                            EXPR_LIDENT@479..480
                                              Lident@479..480 "i"
                                            Comma@480..481 ","
                                            Whitespace@481..482 " "
                                          ARG@482..493
                                            EXPR_BINARY@482..493
                                              EXPR_LIDENT@482..490
                                                Lident@482..489 "current"
                                                Whitespace@489..490 " "
                                              Plus@490..491 "+"
                                              Whitespace@491..492 " "
                                              EXPR_INT@492..493
                                                Int@492..493 "1"
                                          RParen@493..494 ")"
                                          Whitespace@494..495 " "
                                    InKeyword@495..497 "in"
                                    Whitespace@497..502 "\n    "
                                    EXPR_LET_BODY@502..668
                                      EXPR_LET@502..668
                                        LetKeyword@502..505 "let"
                                        Whitespace@505..506 " "
                                        PATTERN_VARIABLE@506..514
                                          Lident@506..513 "add_now"
                                          Whitespace@513..514 " "
                                        Eq@514..515 "="
                                        Whitespace@515..516 " "
                                        EXPR_LET_VALUE@516..533
                                          EXPR_CALL@516..533
                                            EXPR_LIDENT@516..523
                                              Lident@516..523 "ref_get"
                                            ARG_LIST@523..533
                                              LParen@523..524 "("
                                              ARG@524..531
                                                EXPR_LIDENT@524..531
                                                  Lident@524..531 "is_even"
                                              RParen@531..532 ")"
                                              Whitespace@532..533 " "
                                        InKeyword@533..535 "in"
                                        Whitespace@535..540 "\n    "
                                        EXPR_LET_BODY@540..668
                                          EXPR_LET@540..668
                                            LetKeyword@540..543 "let"
                                            Whitespace@543..544 " "
                                            PATTERN_WILDCARD@544..546
                                              WildcardKeyword@544..545 "_"
                                              Whitespace@545..546 " "
                                            Eq@546..547 "="
                                            Whitespace@547..548 " "
                                            EXPR_LET_VALUE@548..575
                                              EXPR_CALL@548..575
                                                EXPR_LIDENT@548..555
                                                  Lident@548..555 "ref_set"
                                                ARG_LIST@555..575
                                                  LParen@555..556 "("
                                                  ARG@556..565
                                                    EXPR_LIDENT@556..563
                                                      Lident@556..563 "is_even"
                                                    Comma@563..564 ","
                                                    Whitespace@564..565 " "
                                                  ARG@565..573
                                                    EXPR_PREFIX@565..573
                                                      Bang@565..566 "!"
                                                      EXPR_LIDENT@566..573
                                                        Lident@566..573 "add_now"
                                                  RParen@573..574 ")"
                                                  Whitespace@574..575 " "
                                            InKeyword@575..577 "in"
                                            Whitespace@577..582 "\n    "
                                            EXPR_LET_BODY@582..668
                                              EXPR_IF@582..668
                                                IfKeyword@582..584 "if"
                                                Whitespace@584..585 " "
                                                EXPR_IF_COND@585..593
                                                  EXPR_LIDENT@585..593
                                                    Lident@585..592 "add_now"
                                                    Whitespace@592..593 " "
                                                EXPR_IF_THEN@593..644
                                                  BLOCK@593..644
                                                    LBrace@593..594 "{"
                                                    Whitespace@594..601 "\n      "
                                                    EXPR_CALL@601..642
                                                      EXPR_LIDENT@601..608
                                                        Lident@601..608 "ref_set"
                                                      ARG_LIST@608..642
                                                        LParen@608..609 "("
                                                        ARG@609..614
                                                          EXPR_LIDENT@609..612
                                                            Lident@609..612 "acc"
                                                          Comma@612..613 ","
                                                          Whitespace@613..614 " "
                                                        ARG@614..636
                                                          EXPR_BINARY@614..636
                                                            EXPR_CALL@614..627
                                                              EXPR_LIDENT@614..621
                                                                Lident@614..621 "ref_get"
                                                              ARG_LIST@621..627
                                                                LParen@621..622 "("
                                                                ARG@622..625
                                                                  EXPR_LIDENT@622..625
                                                                    Lident@622..625 "acc"
                                                                RParen@625..626 ")"
                                                                Whitespace@626..627 " "
                                                            Plus@627..628 "+"
                                                            Whitespace@628..629 " "
                                                            EXPR_LIDENT@629..636
                                                              Lident@629..636 "current"
                                                        RParen@636..637 ")"
                                                        Whitespace@637..642 "\n    "
                                                    RBrace@642..643 "}"
                                                    Whitespace@643..644 " "
                                                ElseKeyword@644..648 "else"
                                                Whitespace@648..649 " "
                                                EXPR_IF_ELSE@649..668
                                                  BLOCK@649..668
                                                    LBrace@649..650 "{"
                                                    Whitespace@650..657 "\n      "
                                                    EXPR_UNIT@657..664
                                                      LParen@657..658 "("
                                                      RParen@658..659 ")"
                                                      Whitespace@659..664 "\n    "
                                                    RBrace@664..665 "}"
                                                    Whitespace@665..668 "\n  "
                              RBrace@668..669 "}"
                        RParen@669..670 ")"
                        Whitespace@670..671 " "
                    InKeyword@671..673 "in"
                    Whitespace@673..676 "\n  "
                    EXPR_LET_BODY@676..689
                      EXPR_CALL@676..689
                        EXPR_LIDENT@676..683
                          Lident@676..683 "ref_get"
                        ARG_LIST@683..689
                          LParen@683..684 "("
                          ARG@684..687
                            EXPR_LIDENT@684..687
                              Lident@684..687 "acc"
                          RParen@687..688 ")"
                          Whitespace@688..689 "\n"
      RBrace@689..690 "}"
      Whitespace@690..692 "\n\n"
  FN@692..921
    FnKeyword@692..694 "fn"
    Whitespace@694..695 " "
    Lident@695..699 "main"
    PARAM_LIST@699..702
      LParen@699..700 "("
      RParen@700..701 ")"
      Whitespace@701..702 " "
    BLOCK@702..921
      LBrace@702..703 "{"
      Whitespace@703..706 "\n  "
      EXPR_LET@706..919
        LetKeyword@706..709 "let"
        Whitespace@709..710 " "
        PATTERN_VARIABLE@710..716
          Lident@710..715 "first"
          Whitespace@715..716 " "
        Eq@716..717 "="
        Whitespace@717..718 " "
        EXPR_LET_VALUE@718..728
          EXPR_CALL@718..728
            EXPR_LIDENT@718..724
              Lident@718..724 "sum_to"
            ARG_LIST@724..728
              LParen@724..725 "("
              ARG@725..726
                EXPR_INT@725..726
                  Int@725..726 "5"
              RParen@726..727 ")"
              Whitespace@727..728 " "
        InKeyword@728..730 "in"
        Whitespace@730..733 "\n  "
        EXPR_LET_BODY@733..919
          EXPR_LET@733..919
            LetKeyword@733..736 "let"
            Whitespace@736..737 " "
            PATTERN_VARIABLE@737..743
              Lident@737..742 "evens"
              Whitespace@742..743 " "
            Eq@743..744 "="
            Whitespace@744..745 " "
            EXPR_LET_VALUE@745..757
              EXPR_CALL@745..757
                EXPR_LIDENT@745..753
                  Lident@745..753 "sum_even"
                ARG_LIST@753..757
                  LParen@753..754 "("
                  ARG@754..755
                    EXPR_INT@754..755
                      Int@754..755 "6"
                  RParen@755..756 ")"
                  Whitespace@756..757 " "
            InKeyword@757..759 "in"
            Whitespace@759..762 "\n  "
            EXPR_LET_BODY@762..919
              EXPR_LET@762..919
                LetKeyword@762..765 "let"
                Whitespace@765..766 " "
                PATTERN_WILDCARD@766..768
                  WildcardKeyword@766..767 "_"
                  Whitespace@767..768 " "
                Eq@768..769 "="
                Whitespace@769..770 " "
                EXPR_LET_VALUE@770..833
                  EXPR_CALL@770..833
                    EXPR_LIDENT@770..784
                      Lident@770..784 "string_println"
                    ARG_LIST@784..833
                      LParen@784..785 "("
                      ARG@785..831
                        EXPR_CALL@785..831
                          EXPR_LIDENT@785..795
                            Lident@785..795 "string_add"
                          ARG_LIST@795..831
                            LParen@795..796 "("
                            ARG@796..810
                              EXPR_STR@796..808
                                Str@796..808 "\"sum_to(5)=\""
                              Comma@808..809 ","
                              Whitespace@809..810 " "
                            ARG@810..830
                              EXPR_CALL@810..830
                                EXPR_LIDENT@810..823
                                  Lident@810..823 "int_to_string"
                                ARG_LIST@823..830
                                  LParen@823..824 "("
                                  ARG@824..829
                                    EXPR_LIDENT@824..829
                                      Lident@824..829 "first"
                                  RParen@829..830 ")"
                            RParen@830..831 ")"
                      RParen@831..832 ")"
                      Whitespace@832..833 " "
                InKeyword@833..835 "in"
                Whitespace@835..838 "\n  "
                EXPR_LET_BODY@838..919
                  EXPR_LET@838..919
                    LetKeyword@838..841 "let"
                    Whitespace@841..842 " "
                    PATTERN_WILDCARD@842..844
                      WildcardKeyword@842..843 "_"
                      Whitespace@843..844 " "
                    Eq@844..845 "="
                    Whitespace@845..846 " "
                    EXPR_LET_VALUE@846..911
                      EXPR_CALL@846..911
                        EXPR_LIDENT@846..860
                          Lident@846..860 "string_println"
                        ARG_LIST@860..911
                          LParen@860..861 "("
                          ARG@861..909
                            EXPR_CALL@861..909
                              EXPR_LIDENT@861..871
                                Lident@861..871 "string_add"
                              ARG_LIST@871..909
                                LParen@871..872 "("
                                ARG@872..888
                                  EXPR_STR@872..886
                                    Str@872..886 "\"sum_even(6)=\""
                                  Comma@886..887 ","
                                  Whitespace@887..888 " "
                                ARG@888..908
                                  EXPR_CALL@888..908
                                    EXPR_LIDENT@888..901
                                      Lident@888..901 "int_to_string"
                                    ARG_LIST@901..908
                                      LParen@901..902 "("
                                      ARG@902..907
                                        EXPR_LIDENT@902..907
                                          Lident@902..907 "evens"
                                      RParen@907..908 ")"
                                RParen@908..909 ")"
                          RParen@909..910 ")"
                          Whitespace@910..911 " "
                    InKeyword@911..913 "in"
                    Whitespace@913..916 "\n  "
                    EXPR_LET_BODY@916..919
                      EXPR_UNIT@916..919
                        LParen@916..917 "("
                        RParen@917..918 ")"
                        Whitespace@918..919 "\n"
      RBrace@919..920 "}"
      Whitespace@920..921 "\n"