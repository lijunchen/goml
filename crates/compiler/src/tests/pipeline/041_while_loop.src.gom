package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func bool_not(x bool) bool {
    return !x
}

func int_add(x int, y int) int {
    return x + y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

type ref_int_x struct {
    value int
}

func ref__Ref_int(value int) *ref_int_x {
    return &ref_int_x{
        value: value,
    }
}

func ref_get__Ref_int(reference *ref_int_x) int {
    return reference.value
}

func ref_set__Ref_int(reference *ref_int_x, value int) struct{} {
    reference.value = value
    return struct{}{}
}

type ref_bool_x struct {
    value bool
}

func ref__Ref_bool(value bool) *ref_bool_x {
    return &ref_bool_x{
        value: value,
    }
}

func ref_get__Ref_bool(reference *ref_bool_x) bool {
    return reference.value
}

func ref_set__Ref_bool(reference *ref_bool_x, value bool) struct{} {
    reference.value = value
    return struct{}{}
}

type Tuple1_unit struct {
    _0 struct{}
}

func sum_to(limit__0 int) int {
    var ret23 int
    var acc__1 *ref_int_x = ref__Ref_int(0)
    var i__2 *ref_int_x = ref__Ref_int(0)
    var cond24 bool
    for {
        var t9 int = ref_get__Ref_int(i__2)
        cond24 = int_less(t9, limit__0)
        if !cond24 {
            break
        }
        var current__3 int = ref_get__Ref_int(i__2)
        var t11 int = ref_get__Ref_int(acc__1)
        var t10 int = int_add(t11, current__3)
        ref_set__Ref_int(acc__1, t10)
        var t12 int = int_add(current__3, 1)
        ref_set__Ref_int(i__2, t12)
    }
    ret23 = ref_get__Ref_int(acc__1)
    return ret23
}

func sum_even(limit__4 int) int {
    var ret25 int
    var acc__5 *ref_int_x = ref__Ref_int(0)
    var i__6 *ref_int_x = ref__Ref_int(0)
    var is_even__7 *ref_bool_x = ref__Ref_bool(true)
    var cond26 bool
    for {
        var t14 int = ref_get__Ref_int(i__6)
        cond26 = int_less(t14, limit__4)
        if !cond26 {
            break
        }
        var current__8 int = ref_get__Ref_int(i__6)
        var t15 int = int_add(current__8, 1)
        ref_set__Ref_int(i__6, t15)
        var add_now__9 bool = ref_get__Ref_bool(is_even__7)
        var t16 bool = bool_not(add_now__9)
        ref_set__Ref_bool(is_even__7, t16)
        if add_now__9 {
            var t18 int = ref_get__Ref_int(acc__5)
            var t17 int = int_add(t18, current__8)
            ref_set__Ref_int(acc__5, t17)
        } else {}
    }
    ret25 = ref_get__Ref_int(acc__5)
    return ret25
}

func main0() struct{} {
    var ret27 struct{}
    var first__10 int = sum_to(5)
    var evens__11 int = sum_even(6)
    var t20 string = int_to_string(first__10)
    var t19 string = string_add("sum_to(5)=", t20)
    string_println(t19)
    var t22 string = int_to_string(evens__11)
    var t21 string = string_add("sum_even(6)=", t22)
    string_println(t21)
    ret27 = struct{}{}
    return ret27
}

func main() {
    main0()
}
