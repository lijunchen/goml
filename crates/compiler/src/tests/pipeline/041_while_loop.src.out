sum_to(5)=10
sum_even(6)=6
